"use strict";(self.webpackChunkzio_site=self.webpackChunkzio_site||[]).push([[80059],{3905:(e,n,t)=>{t.d(n,{Zo:()=>u,kt:()=>m});var r=t(67294);function a(e,n,t){return n in e?Object.defineProperty(e,n,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[n]=t,e}function o(e,n){var t=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);n&&(r=r.filter((function(n){return Object.getOwnPropertyDescriptor(e,n).enumerable}))),t.push.apply(t,r)}return t}function i(e){for(var n=1;n<arguments.length;n++){var t=null!=arguments[n]?arguments[n]:{};n%2?o(Object(t),!0).forEach((function(n){a(e,n,t[n])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(t)):o(Object(t)).forEach((function(n){Object.defineProperty(e,n,Object.getOwnPropertyDescriptor(t,n))}))}return e}function c(e,n){if(null==e)return{};var t,r,a=function(e,n){if(null==e)return{};var t,r,a={},o=Object.keys(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||(a[t]=e[t]);return a}(e,n);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(a[t]=e[t])}return a}var l=r.createContext({}),s=function(e){var n=r.useContext(l),t=n;return e&&(t="function"==typeof e?e(n):i(i({},n),e)),t},u=function(e){var n=s(e.components);return r.createElement(l.Provider,{value:n},e.children)},p="mdxType",f={inlineCode:"code",wrapper:function(e){var n=e.children;return r.createElement(r.Fragment,{},n)}},d=r.forwardRef((function(e,n){var t=e.components,a=e.mdxType,o=e.originalType,l=e.parentName,u=c(e,["components","mdxType","originalType","parentName"]),p=s(t),d=a,m=p["".concat(l,".").concat(d)]||p[d]||f[d]||o;return t?r.createElement(m,i(i({ref:n},u),{},{components:t})):r.createElement(m,i({ref:n},u))}));function m(e,n){var t=arguments,a=n&&n.mdxType;if("string"==typeof e||a){var o=t.length,i=new Array(o);i[0]=d;var c={};for(var l in n)hasOwnProperty.call(n,l)&&(c[l]=n[l]);c.originalType=e,c[p]="string"==typeof e?e:a,i[1]=c;for(var s=2;s<o;s++)i[s]=t[s];return r.createElement.apply(null,i)}return r.createElement.apply(null,t)}d.displayName="MDXCreateElement"},82523:(e,n,t)=>{t.r(n),t.d(n,{assets:()=>l,contentTitle:()=>i,default:()=>f,frontMatter:()=>o,metadata:()=>c,toc:()=>s});var r=t(87462),a=(t(67294),t(3905));const o={id:"sandboxing",title:"Sandboxing",sidebar_label:"6. Sandboxing"},i=void 0,c={unversionedId:"reference/error-management/recovering/sandboxing",id:"reference/error-management/recovering/sandboxing",title:"Sandboxing",description:"We know that a ZIO effect may fail due to a failure, a defect, a fiber interruption, or a combination of these causes. So a ZIO effect may contain more than one cause. Using the ZIO#sandbox operator, we can sandbox all errors of a ZIO application, whether the cause is a failure, defect, or a fiber interruption or combination of these. This operator exposes the full cause of a ZIO effect into the error channel:",source:"@site/docs/reference/error-management/recovering/sandboxing.md",sourceDirName:"reference/error-management/recovering",slug:"/reference/error-management/recovering/sandboxing",permalink:"/reference/error-management/recovering/sandboxing",draft:!1,editUrl:"https://github.com/zio/zio/edit/series/2.x/docs/reference/error-management/recovering/sandboxing.md",tags:[],version:"current",frontMatter:{id:"sandboxing",title:"Sandboxing",sidebar_label:"6. Sandboxing"},sidebar:"reference-sidebar",previous:{title:"5. Timing out",permalink:"/reference/error-management/recovering/timing-out"},next:{title:"Error Accumulation",permalink:"/reference/error-management/error-accumulation"}},l={},s=[],u={toc:s},p="wrapper";function f(e){let{components:n,...t}=e;return(0,a.kt)(p,(0,r.Z)({},u,t,{components:n,mdxType:"MDXLayout"}),(0,a.kt)("p",null,"We know that a ZIO effect may fail due to a failure, a defect, a fiber interruption, or a combination of these causes. So a ZIO effect may contain more than one cause. Using the ",(0,a.kt)("inlineCode",{parentName:"p"},"ZIO#sandbox")," operator, we can sandbox all errors of a ZIO application, whether the cause is a failure, defect, or a fiber interruption or combination of these. This operator exposes the full cause of a ZIO effect into the error channel:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-scala"},"trait ZIO[-R, +E, +A] {\n  def sandbox: ZIO[R, Cause[E], A]\n}\n")),(0,a.kt)("p",null,"We can use the ",(0,a.kt)("inlineCode",{parentName:"p"},"ZIO#sandbox")," operator to uncover the full causes of an ",(0,a.kt)("em",{parentName:"p"},"exceptional effect"),". So we can see all the errors that occurred as a type of ",(0,a.kt)("inlineCode",{parentName:"p"},"Cause[E]")," at the error channel of the ",(0,a.kt)("inlineCode",{parentName:"p"},"ZIO")," data type. So then we can use normal error-handling operators such as ",(0,a.kt)("inlineCode",{parentName:"p"},"ZIO#catchSome")," and ",(0,a.kt)("inlineCode",{parentName:"p"},"ZIO#catchAll")," operators:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-scala"},'import zio._\n\nobject MainApp extends ZIOAppDefault {\n  val effect: ZIO[Any, String, String] =\n    ZIO.succeed("primary result") *> ZIO.fail("Oh uh!")\n\n  val myApp: ZIO[Any, Cause[String], String] =\n    effect.sandbox.catchSome {\n      case Cause.Interrupt(fiberId, _) =>\n        ZIO.debug(s"Caught interruption of a fiber with id: $fiberId") *>\n          ZIO.succeed("fallback result on fiber interruption")\n      case Cause.Die(value, _) =>\n        ZIO.debug(s"Caught a defect: $value") *>\n          ZIO.succeed("fallback result on defect")\n      case Cause.Fail(value, _) =>\n        ZIO.debug(s"Caught a failure: $value") *>\n          ZIO.succeed("fallback result on failure")\n    }\n\n  val finalApp: ZIO[Any, String, String] = myApp.unsandbox.debug("final result")\n\n  def run = finalApp\n}\n\n// Output:\n// Caught a failure: Oh uh!\n// final result: fallback result on failure\n')),(0,a.kt)("p",null,"Using the ",(0,a.kt)("inlineCode",{parentName:"p"},"sandbox")," operation we are exposing the full cause of an effect. So then we have access to the underlying cause in more detail. After handling exposed causes using ",(0,a.kt)("inlineCode",{parentName:"p"},"ZIO#catch*")," operators, we can undo the ",(0,a.kt)("inlineCode",{parentName:"p"},"sandbox")," operation using the ",(0,a.kt)("inlineCode",{parentName:"p"},"unsandbox")," operation. It will submerge the full cause (",(0,a.kt)("inlineCode",{parentName:"p"},"Cause[E]"),") again:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-scala"},'import zio._\n\nval effect: ZIO[Any, String, String] =\n  ZIO.succeed("primary result") *> ZIO.fail("Oh uh!")\n\neffect            // ZIO[Any, String, String]\n  .sandbox        // ZIO[Any, Cause[String], String]\n  .catchSome(???) // ZIO[Any, Cause[String], String]\n  .unsandbox      // ZIO[Any, String, String]\n')),(0,a.kt)("p",null,"There is another version of sandbox called ",(0,a.kt)("inlineCode",{parentName:"p"},"ZIO#sandboxWith"),". This operator helps us to sandbox, then catch all causes, and then unsandbox back:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-scala"},"trait ZIO[-R, +E, +A] {\n  def sandboxWith[R1 <: R, E2, B](f: ZIO[R1, Cause[E], A] => ZIO[R1, Cause[E2], B])\n}\n")),(0,a.kt)("p",null,"Let's try the previous example using this operator:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-scala"},'import zio._\n\nobject MainApp extends ZIOAppDefault {\n  val effect: ZIO[Any, String, String] =\n    ZIO.succeed("primary result") *> ZIO.fail("Oh uh!")\n\n  val myApp =\n    effect.sandboxWith[Any, String, String] { e =>\n      e.catchSome {\n        case Cause.Interrupt(fiberId, _) =>\n          ZIO.debug(s"Caught interruption of a fiber with id: $fiberId") *>\n            ZIO.succeed("fallback result on fiber interruption")\n        case Cause.Die(value, _) =>\n          ZIO.debug(s"Caught a defect: $value") *>\n            ZIO.succeed("fallback result on defect")\n        case Cause.Fail(value, _) =>\n          ZIO.debug(s"Caught a failure: $value") *>\n            ZIO.succeed("fallback result on failure")\n      }\n    }\n  def run = myApp.debug\n}\n')))}f.isMDXComponent=!0}}]);